`type-annotation-spacing` and `key-spacing`: conflict

### Validations

- [X] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Describe the bug

`key-spacing` should only check for key values not for types but `key-spacing` is also checking types. Types should only be checked by `type-annotation-spacing` and not `key-spacing`.

This is my current `eslint.config.js`:
```js
// eslint.config.js
import parserTs from '@typescript-eslint/parser'
import stylistic from '@stylistic/eslint-plugin'

export default [
  {
    files: ["**/*.ts"],
    plugins: {
      '@stylistic': stylistic
    },
    languageOptions: {
      parser: parserTs
    },
    rules: {
      '@stylistic/type-annotation-spacing': ['error', {
        "before": true,
        "after": true
      }],
      '@stylistic/key-spacing': ['error', {
        "beforeColon": false,
        "afterColon": true,
        "mode": "strict"
      }]
    }
  }
];
```

As an example, this is how I want my code to be formatted:
```ts
// deno-lint-ignore-file
let fooBar : string = "bar";
//        ^ ^ 1 space

function foo() : {e : string} {return {e: ''};}
//            ^ ^  ^ ^ 1 space          ^ No space

class Foo {
    name : string | undefined;
//      ^ ^ 1 space
}

type Fooo = () => {};
```

```
  4:19  error  Extra space after key 'e'     @stylistic/key-spacing
  7:5   error  Extra space after key 'name'  @stylistic/key-spacing

✖ 2 problems (2 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.
```

The rules described conflict.

### Reproduction

https://github.com/DenoPlayground/eslint-example

https://github.com/eslint/eslint/issues/18701

### Contributes

- [ ] I am willing to submit a PR to fix this issue
- [ ] I am willing to submit a PR with failing tests

--- a/packages/eslint-plugin/rules/indent/types.d.ts
+++ b/packages/eslint-plugin/rules/indent/types.d.ts
@@ -1,6 +1,6 @@
 /* GENERATED, DO NOT EDIT DIRECTLY */
 
-/* @checksum: zdiMvtTqyx */
+/* @checksum: B8DyPsI0dM */
 
 export type IndentSchema0 = 'tab' | number
 

--- a/packages/eslint-plugin/rules/key-spacing/README._js_.md
+++ b/packages/eslint-plugin/rules/key-spacing/README._js_.md
@@ -28,6 +28,7 @@ This rule has an object option:
   - `"value"`: enforces horizontal alignment of values in object literals.
   - `"colon"` enforces horizontal alignment of both colons and values in object literals.
 - `"align"` with an object value allows for fine-grained spacing when values are being aligned in object literals.
+- `"ignoredNodes"` can be used to disable spacing checking for the AST node. It can be used to resolve conflicts with [`type-annotation-spacing`](../ts/type-annotation-spacing).
 - `"singleLine"` specifies a spacing style for single-line object literals.
 - `"multiLine"` specifies a spacing style for multi-line object literals.
 
@@ -303,6 +304,33 @@ var obj = {
     "one":  1,
     "seven":7
 }
+```
+
+:::
+
+:::
+
+### ignoredNodes
+
+The "ignoredNodes" object property is optional (default: []).
+
+Examples of **correct** code for this rule with sample `{ "ignoredNodes": [] }` options:
+
+::: correct
+
+```js
+/*eslint "key-spacing": [2, { "ignoredNodes": ["ObjectExpression"] }]*/
+var obj = {
+    a: 1,
+    b : 2,
+    c :3,
+}
+/*eslint "key-spacing": [2, { "ignoredNodes": ["ObjectPattern"] }]*/
+var {
+    a: b,
+    c : d,
+    e :f,
+} = obj
 ```
 
 :::

--- a/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.test.ts
+++ b/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.test.ts
@@ -1251,6 +1251,82 @@ run({
       `,
       options: [{ align: 'colon' }],
     },
+    {
+      code: $`
+        var a = { key : value };
+      `,
+      options: [{ ignoredNodes: ['Property'] }],
+    },
+    {
+      code: $`
+        import "./foo" with { type : "json" }
+      `,
+      options: [{ ignoredNodes: ['ImportAttribute'] }],
+    },
+    {
+      code: $`
+        export {foo} from "./foo" with { type : "json" }
+      `,
+      options: [{ ignoredNodes: ['ImportAttribute'] }],
+    },
+    {
+      code: $`
+        var obj = {
+          'a': (42 - 12),
+          foobar: 'value',
+          [(expr)]: val
+        }
+      `,
+      options: [{
+        align: 'colon',
+        ignoredNodes: ['ObjectExpression'],
+      }],
+    },
+    {
+      code: $`
+        import foo from "./foo" with
+            {
+              type : "json",
+              foo : "bar"
+            }
+      `,
+      options: [{
+        beforeColon: true,
+        afterColon: true,
+        align: 'colon',
+        ignoredNodes: ['ImportDeclaration'],
+      }],
+    },
+    {
+      code: $`
+        export {foo} from "./foo" with
+            {
+              type : "json",
+              foo : "bar"
+            }
+      `,
+      options: [{
+        beforeColon: true,
+        afterColon: true,
+        align: 'colon',
+        ignoredNodes: ['ExportNamedDeclaration'],
+      }],
+    },
+    {
+      code: $`
+        export * from "./foo" with
+            {
+              type : "json",
+              foo : "bar"
+            }
+      `,
+      options: [{
+        beforeColon: true,
+        afterColon: true,
+        align: 'colon',
+        ignoredNodes: ['ExportAllDeclaration'],
+      }],
+    },
   ],
   invalid: [
     {

--- a/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.test.ts
+++ b/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.test.ts
@@ -1273,6 +1273,36 @@ run({
     },
     {
       code: $`
+        import foo from "./foo" with
+            {
+              type : "json",
+              foo : "bar"
+            }
+      `,
+      options: [{ ignoredNodes: ['ImportDeclaration'] }],
+    },
+    {
+      code: $`
+        export {foo} from "./foo" with
+            {
+              type : "json",
+              foo : "bar"
+            }
+      `,
+      options: [{ ignoredNodes: ['ExportNamedDeclaration'] }],
+    },
+    {
+      code: $`
+        export * from "./foo" with
+            {
+              type : "json",
+              foo : "bar"
+            }
+      `,
+      options: [{ ignoredNodes: ['ExportAllDeclaration'] }],
+    },
+    {
+      code: $`
         var obj = {
           'a': (42 - 12),
           foobar: 'value',
@@ -1293,8 +1323,6 @@ run({
             }
       `,
       options: [{
-        beforeColon: true,
-        afterColon: true,
         align: 'colon',
         ignoredNodes: ['ImportDeclaration'],
       }],
@@ -1308,8 +1336,6 @@ run({
             }
       `,
       options: [{
-        beforeColon: true,
-        afterColon: true,
         align: 'colon',
         ignoredNodes: ['ExportNamedDeclaration'],
       }],
@@ -1323,8 +1349,6 @@ run({
             }
       `,
       options: [{
-        beforeColon: true,
-        afterColon: true,
         align: 'colon',
         ignoredNodes: ['ExportAllDeclaration'],
       }],

--- a/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.test.ts
+++ b/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.test.ts
@@ -1253,21 +1253,23 @@ run({
     },
     {
       code: $`
-        var a = { key : value };
+        var obj = {
+          'a': 42 - 12,
+          foobar : 'value',
+          [(expr)] :val
+        }
       `,
-      options: [{ ignoredNodes: ['Property'] }],
+      options: [{ ignoredNodes: ['ObjectExpression'] }],
     },
     {
       code: $`
-        import "./foo" with { type : "json" }
+        var {
+          a: b,
+          c : d,
+          e :f,
+        } = obj
       `,
-      options: [{ ignoredNodes: ['ImportAttribute'] }],
-    },
-    {
-      code: $`
-        export {foo} from "./foo" with { type : "json" }
-      `,
-      options: [{ ignoredNodes: ['ImportAttribute'] }],
+      options: [{ ignoredNodes: ['ObjectPattern'] }],
     },
     {
       code: $`

--- a/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.ts
+++ b/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.ts
@@ -10,9 +10,8 @@ import { createRule } from '#utils/create-rule'
 import { getStringLength } from '#utils/string'
 
 const listeningNodes = [
-  'Property',
-  'ImportAttribute',
   'ObjectExpression',
+  'ObjectPattern',
   'ImportDeclaration',
   'ExportNamedDeclaration',
   'ExportAllDeclaration',
@@ -663,7 +662,7 @@ export default createRule<RuleOptions, MessageIds>({
      * @param lineOptions Configured singleLine or multiLine options
      */
     function verifySpacing(node: Tree.Property | Tree.ImportAttribute, lineOptions: { beforeColon: number, afterColon: number, mode: 'strict' | 'minimum' }) {
-      if (ignoredNodes.includes(node.type))
+      if (ignoredNodes.includes(node.parent.type))
         return
       const actual = getPropertyWhitespace(node)
 
@@ -742,7 +741,7 @@ export default createRule<RuleOptions, MessageIds>({
         verifySpacing(node, isSingleLine(node.parent) ? singleLineOptions : multiLineOptions)
       },
       ImportAttribute(node) {
-        const parent = node.parent as Tree.ImportDeclaration | Tree.ExportNamedDeclaration | Tree.ExportAllDeclaration
+        const parent = node.parent
         verifySpacing(node, isSingleLineImportAttributes(parent, sourceCode) ? singleLineOptions : multiLineOptions)
       },
     }

--- a/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.ts
+++ b/packages/eslint-plugin/rules/key-spacing/key-spacing._js_.ts
@@ -9,6 +9,19 @@ import { getStaticPropertyName, isClosingBraceToken, isColonToken, isOpeningBrac
 import { createRule } from '#utils/create-rule'
 import { getStringLength } from '#utils/string'
 
+const listeningNodes = [
+  'Property',
+  'ImportAttribute',
+  'ObjectExpression',
+  'ImportDeclaration',
+  'ExportNamedDeclaration',
+  'ExportAllDeclaration',
+
+  'TSTypeLiteral',
+  'TSInterfaceBody',
+  'ClassBody',
+] satisfies (keyof typeof Tree.AST_NODE_TYPES)[]
+
 /**
  * Checks whether a string contains a line terminator as defined in
  * http://www.ecma-international.org/ecma-262/5.1/#sec-7.3
@@ -132,6 +145,7 @@ function initOptions(toOptions: any, fromOptions: any) {
       }
     }
   }
+  toOptions.ignoredNodes = fromOptions.ignoredNodes || []
 
   return toOptions
 }
@@ -190,6 +204,13 @@ export default createRule<RuleOptions, MessageIds>({
             },
             afterColon: {
               type: 'boolean',
+            },
+            ignoredNodes: {
+              type: 'array',
+              items: {
+                type: 'string',
+                enum: listeningNodes,
+              },
             },
           },
           additionalProperties: false,
@@ -342,6 +363,7 @@ export default createRule<RuleOptions, MessageIds>({
     const multiLineOptions = ruleOptions.multiLine
     const singleLineOptions = ruleOptions.singleLine
     const alignmentOptions = ruleOptions.align || null
+    const ignoredNodes: Tree.AST_NODE_TYPES[] = ruleOptions.ignoredNodes
 
     const sourceCode = context.sourceCode
 
@@ -641,6 +663,8 @@ export default createRule<RuleOptions, MessageIds>({
      * @param lineOptions Configured singleLine or multiLine options
      */
     function verifySpacing(node: Tree.Property | Tree.ImportAttribute, lineOptions: { beforeColon: number, afterColon: number, mode: 'strict' | 'minimum' }) {
+      if (ignoredNodes.includes(node.type))
+        return
       const actual = getPropertyWhitespace(node)
 
       if (actual) { // Object literal getters/setters lack colons
@@ -677,6 +701,8 @@ export default createRule<RuleOptions, MessageIds>({
     }
 
     function verifyImportAttributes(node: Tree.ImportDeclaration | Tree.ExportNamedDeclaration | Tree.ExportAllDeclaration) {
+      if (ignoredNodes.includes(node.type))
+        return
       if (!node.attributes)
         // The old parser's AST does not have attributes.
         return
@@ -691,6 +717,8 @@ export default createRule<RuleOptions, MessageIds>({
     if (alignmentOptions) { // Verify vertical alignment
       return {
         ObjectExpression(node) {
+          if (ignoredNodes.includes(node.type))
+            return
           if (isSingleLine(node))
             verifyListSpacing(node.properties.filter(isKeyValueProperty), singleLineOptions)
           else

--- a/packages/eslint-plugin/rules/key-spacing/key-spacing._ts_.test.ts
+++ b/packages/eslint-plugin/rules/key-spacing/key-spacing._ts_.test.ts
@@ -616,6 +616,33 @@ run({
         }
       `,
     },
+    {
+      code: $`
+        type X = {
+          a: number;
+          abc: string
+        };
+      `,
+      options: [{ align: 'value', ignoredNodes: ['TSTypeLiteral'] }],
+    },
+    {
+      code: $`
+        interface X {
+          a: number;
+          abc: string
+        };
+      `,
+      options: [{ align: 'value', ignoredNodes: ['TSInterfaceBody'] }],
+    },
+    {
+      code: $`
+        class X {
+          a: number;
+          abc: string
+        };
+      `,
+      options: [{ align: 'value', ignoredNodes: ['ClassBody'] }],
+    },
   ],
   invalid: [
     // align: value

--- a/packages/eslint-plugin/rules/key-spacing/key-spacing._ts_.test.ts
+++ b/packages/eslint-plugin/rules/key-spacing/key-spacing._ts_.test.ts
@@ -619,6 +619,36 @@ run({
     {
       code: $`
         type X = {
+          a :number;
+          b: string;
+          c :string;
+        };
+      `,
+      options: [{ ignoredNodes: ['TSTypeLiteral'] }],
+    },
+    {
+      code: $`
+        interface X {
+          a :number;
+          b: string;
+          c :string;
+        };
+      `,
+      options: [{ ignoredNodes: ['TSInterfaceBody'] }],
+    },
+    {
+      code: $`
+        class X {
+          a :number;
+          b: string;
+          c :string;
+        };
+      `,
+      options: [{ ignoredNodes: ['ClassBody'] }],
+    },
+    {
+      code: $`
+        type X = {
           a: number;
           abc: string
         };

--- a/packages/eslint-plugin/rules/key-spacing/key-spacing._ts_.ts
+++ b/packages/eslint-plugin/rules/key-spacing/key-spacing._ts_.ts
@@ -38,6 +38,7 @@ export default createRule<RuleOptions, MessageIds>({
   defaultOptions: [{}],
   create(context, [_options]) {
     const options: OptionsUnion = _options || {}
+    const ignoredNodes = options.ignoredNodes || []
     const sourceCode = context.sourceCode
     const baseRules = baseRule.create(context)
 
@@ -370,6 +371,9 @@ export default createRule<RuleOptions, MessageIds>({
         | Tree.TSInterfaceBody
         | Tree.TSTypeLiteral,
     ): void {
+      if (ignoredNodes.includes(body.type))
+        return
+
       const isSingleLine = body.loc.start.line === body.loc.end.line
 
       const members = body.type === AST_NODE_TYPES.TSTypeLiteral

--- a/packages/eslint-plugin/rules/key-spacing/types.d.ts
+++ b/packages/eslint-plugin/rules/key-spacing/types.d.ts
@@ -1,6 +1,6 @@
 /* GENERATED, DO NOT EDIT DIRECTLY */
 
-/* @checksum: En2BtZQg2T */
+/* @checksum: cWqWPSqo6p */
 
 export type KeySpacingSchema0 =
   | {
@@ -15,6 +15,17 @@ export type KeySpacingSchema0 =
     mode?: 'strict' | 'minimum'
     beforeColon?: boolean
     afterColon?: boolean
+    ignoredNodes?: (
+      | 'Property'
+      | 'ImportAttribute'
+      | 'ObjectExpression'
+      | 'ImportDeclaration'
+      | 'ExportNamedDeclaration'
+      | 'ExportAllDeclaration'
+      | 'TSTypeLiteral'
+      | 'TSInterfaceBody'
+      | 'ClassBody'
+    )[]
   }
   | {
     singleLine?: {

--- a/packages/eslint-plugin/rules/key-spacing/types.d.ts
+++ b/packages/eslint-plugin/rules/key-spacing/types.d.ts
@@ -1,6 +1,6 @@
 /* GENERATED, DO NOT EDIT DIRECTLY */
 
-/* @checksum: cWqWPSqo6p */
+/* @checksum: Vl4RUSFlaP */
 
 export type KeySpacingSchema0 =
   | {
@@ -16,9 +16,8 @@ export type KeySpacingSchema0 =
     beforeColon?: boolean
     afterColon?: boolean
     ignoredNodes?: (
-      | 'Property'
-      | 'ImportAttribute'
       | 'ObjectExpression'
+      | 'ObjectPattern'
       | 'ImportDeclaration'
       | 'ExportNamedDeclaration'
       | 'ExportAllDeclaration'

